### PR TITLE
Add entries to in-memory store on Put()

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Reference counting for serialized objects doesn't currently work for `ray.put` objects because we don't have in-memory object store entries for them. This PR puts an `InPlasmaError` for those objects, standardizing the semantics for all stored objects (there's always an `InPlasmaError` for those in plasma).

Also adds reference counting tests that check that this is working.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: python/ray/serialization.py [hunk 1]
```diff
@@ -184,6 +184,11 @@ def object_id_deserializer(serialized_obj):
                 # UniqueIDs are serialized as
                 # (class name, (unique bytes,)).
                 outer_id = context.get_outer_object_id()
+                # outer_id is None in the case that this ObjectID was closed
+                # over in a function or pickled directly using pickle.dumps().
+                # TODO(edoakes): make sure these objects are pinned forever.
+                if outer_id is None:
```
Comment (stephanie-wang) [inline python/ray/serialization.py]: Move this bit to #7358?

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc [hunk 2]
```diff
@@ -212,6 +212,11 @@ Status CoreWorkerMemoryStore::Put(const RayObject &object, const ObjectID &objec
     }
   }
 
+  // Must be called without holding the lock.
```
Comment (stephanie-wang) [inline src/ray/core_worker/store_provider/memory_store/memory_store.cc]: Can you expand this comment a bit to explain the potential deadlock?

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc [hunk 3]
```diff
@@ -250,7 +255,6 @@ Status CoreWorkerMemoryStore::GetImpl(const std::vector<ObjectID> &object_ids,
       if (iter != objects_.end()) {
         (*results)[i] = iter->second;
         if (remove_after_get) {
-          RAY_LOG(ERROR) << "REMOVE_AFTER_GET";
```
Comment (stephanie-wang) [inline src/ray/core_worker/store_provider/memory_store/memory_store.cc]: Oops, I think this was me :)

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc [hunk 4]
```diff
@@ -437,7 +441,6 @@ bool CoreWorkerMemoryStore::Contains(const ObjectID &object_id, bool *in_plasma)
   if (it != objects_.end()) {
     if (it->second->IsInPlasmaError()) {
       *in_plasma = true;
-      return false;
```
Comment (stephanie-wang) [inline src/ray/core_worker/store_provider/memory_store/memory_store.cc]: Is it okay to remove this?

**Review by stephanie-wang — APPROVED** [review]
So much nicer!

File: python/ray/serialization.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serialization.py]: Ah yeah #7358 will fix it but this prevents a nasty segfault. If the other gets merged first I'll remove this.

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc (hunk 4 above)
Comment (edoakes) [reply src/ray/core_worker/store_provider/memory_store/memory_store.cc]: Yep! Just made the API a bit more sane (updated header and callers)

**Pull request merged**